interp/interp_test.go

```diff
@@ -191,3 +190,341 @@ func TestPiecewiseConstantPredict(t *testing.T) {
 	leftYs := []float64{-0.5, 1.5, 1.5, 1.5, 1, 1}
 	testInterpolatorPredict(t, pc, testXs, leftYs, 0)
 }
+
+func TestPiecewiseCubic(t *testing.T) {
+	t.Parallel()
+	const (
+		h        = 1e-8
+		valueTol = 1e-13
+		derivTol = 1e-6
+	)
+	for i, test := range []struct {
+		xs []float64
+		f  func(float64) float64
+		df func(float64) float64
+	}{
+		{
+			xs: []float64{-1.001, 0.2, 2},
+			f:  func(x float64) float64 { return x * x },
+			df: func(x float64) float64 { return 2 * x },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 2},
+			f:  func(x float64) float64 { return x * x },
+			df: func(x float64) float64 { return 2 * x },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 10},
+			f:  func(x float64) float64 { return 1.5*x - 1 },
+			df: func(x float64) float64 { return 1.5 },
+		},
+		{
+			xs: []float64{-1.001, 0.2, 10},
+			f:  func(x float64) float64 { return -1 },
+			df: func(x float64) float64 { return 0 },
+		},
+		{
+			xs: []float64{-1.1, 0.2, 0.99, 2.5, 2.99},
+			f:  math.Cos,
+			df: math.Sin,
+		},
+		{
+			xs: []float64{-1.1, 0.2, 0.99, 2.5, 2.99},
+			f:  math.Exp,
+			df: math.Exp,
+		},
+		{
+			xs: []float64{-1.1, 0.2, 0.99, 2.5, 2.99},
+			f:  func(x float64) float64 { return math.Sin(x * x) },
+			df: func(x float64) float64 { return -2 * x * math.Cos(x*x) },
+		},
+	} {
+		ys := applyFunc(test.xs, test.f)
+		dydxs := applyFunc(test.xs, test.df)
+		var pc PiecewiseCubic
+		pc.fitWithDerivatives(test.xs, ys, dydxs)
+		n := len(test.xs)
+		for j := 0; j < n; j++ {
+			x := test.xs[j]
+			got := pc.Predict(x)
+			want := test.f(x)
+			if math.Abs(got-want) > valueTol {
+				t.Errorf("Mismatch in interpolated value at x == %g for test case %d: got %v, want %g", x, i, got, want)
+			}
+			if j < n-1 {
+				got = pc.coeffs.At(j, 0)
+				if math.Abs(got-want) > valueTol {
+					t.Errorf("Mismatch in 0-th order interpolation coefficient in %d-th node for test case %d: got %v, want %g", j, i, got, want)
```

Comment:
I don't think it's necessary to check values of the internal coefficients. Removing it would simplify the test a lot.

Reply:
I disagree. I think it gives additional security that the implementation is correct.

Reply:
Hmm, OK.